File: codetandem/secrets.py
"""Secure API key storage using system keyring."""

import keyring
import keyring.errors
from typing import Optional


SERVICE_NAME = "codetandem"


def set_api_key(provider: str, api_key: str):
    """
    Store an API key securely in the system keyring.

    Args:
        provider: The AI provider name (e.g., 'openai', 'anthropic')
        api_key: The API key to store
    """
    username = f"{provider}_api_key"
    keyring.set_password(SERVICE_NAME, username, api_key)


def get_api_key(provider: str) -> Optional[str]:
    """
    Retrieve an API key from the system keyring.

    Args:
        provider: The AI provider name (e.g., 'openai', 'anthropic')

    Returns:
        The API key or None if not found
    """
    username = f"{provider}_api_key"
    return keyring.get_password(SERVICE_NAME, username)


def delete_api_key(provider: str):
    """
    Delete an API key from the system keyring.

    Args:
        provider: The AI provider name (e.g., 'openai', 'anthropic')
    """
    username = f"{provider}_api_key"
    try:
        keyring.delete_password(SERVICE_NAME, username)
    except keyring.errors.PasswordDeleteError:
        # Key doesn't exist, which is fine
        pass

File: src/utils/secrets.ts
/**
 * Secure API key storage using system keyring (keytar)
 */

import keytar from 'keytar';

const SERVICE_NAME = 'codetandem';

/**
 * Store an API key securely in the system keyring
 */
export async function setApiKey(provider: string, apiKey: string): Promise<void> {
  const username = `${provider}_api_key`;
  await keytar.setPassword(SERVICE_NAME, username, apiKey);
}

/**
 * Retrieve an API key from the system keyring
 */
export async function getApiKey(provider: string): Promise<string | null> {
  const username = `${provider}_api_key`;
  return await keytar.getPassword(SERVICE_NAME, username);
}

/**
 * Delete an API key from the system keyring
 */
export async function deleteApiKey(provider: string): Promise<boolean> {
  const username = `${provider}_api_key`;
  return await keytar.deletePassword(SERVICE_NAME, username);
}

/**
 * List all stored API keys (returns provider names only)
 */
export async function listApiKeys(): Promise<string[]> {
  const credentials = await keytar.findCredentials(SERVICE_NAME);
  return credentials
    .map((cred) => cred.account.replace('_api_key', ''))
    .filter((provider) => provider.length > 0);
}
